### ISSUE TITLE
Switch CLIs from argparse to click

### ISSUE THREAD

**Issue by bwoodsend** [pyinstaller/pyinstaller#5559]
Transitioning to click will:

- Make adding/changing options much easier.
- Allow us to automatically enable bash/fish/zsh autocompletion which, given how many options PyInstaller has and how inconsistent they are with regards to word deliminators and pluralisation (is it `--add-data or ``-adddatas`?), would be a really nice feature.
- Allow us to jettison some low level clutter code such as [_SmartFormatter](https://github.com/pyinstaller/pyinstaller/blob/c24117a98f4322de861f62019fb3c3c869a85245/PyInstaller/__main__.py#L37) which deals with word wrapping.

We have quite a few CLIs but the two big ones are `PyInstaller/__main__.py` and `PyInstaller/utils/cliutils.makespec.py`.

**Comment by johnthagen** [comment]
I would recommend you also consider Typer, which is built on to of Click: https://github.com/tiangolo/typer

**Comment by bwoodsend** [comment]
Actually we (mostly me but Riz too) have changed our minds about click. It forces you to make your argument parsing CLI-ified function the very top level of your code which:

1. Requires some quite extensive ripping up of PyInstaller's floorboards (which needs doing anyway so this isn't too bad).
2. Makes it very hard to pass shared arguments between multiple similar CLIs. We got this mostly working but the resultant code was incomprehensible and brittle as hell.
3. Makes it impossible to test. I want to be able to put `assert parse_args("--pathex", "/foo/:/bar", "--pathex=/baz")["pathex"] == ["/foo", "/bar", "/baz"]` so that we can fully test the more customised (and possibly platform dependent) corners of our CLI in under a second rather than indirectly testing half our CLI in 8 hours like we do currently.

With argument parsing inseparably coupled with running the application and calling `sys.exit()` (which IMO no API should ever do implicitly) I'd say that click is practically useless to us. We'll keep the branches around but we won't be using them unless someone finds a way to decouple the 3.

**Comment by johnthagen** [comment]
Yeah, not trying to prescribe anything (more of a drive by comment). We've personally had good success with Typer as a much easier to work with skin over Click.

> Makes it very hard to pass shared arguments between multiple similar CLIs. 

This one at least, I think Typer's story is a little better than Click.

Because Typer uses objects in the type hint location, you reuse them like:

```py
import typer

app = typer.Typer()

name_arg = typer.Argument("Bruce", help="Who to greet")

@app.command()
def create(name: str = name_arg):
    typer.echo(f"Creating item: {item}")


@app.command()
def delete(name: str = name_arg):
    typer.echo(f"Deleting item: {item}")
```

Anyway, I totally agree that Click/Typer do sorta lead you down a specific path for CLI arguments. Not trying to prescribe anything. Just wanted to pass along a cool project we've had a lot of success with.

The type hints also work really well in editors such as PyCharm and VS Code and type checkers such as `mypy`.

**Comment by bwoodsend** [comment]
We're not particularly keen on argparse either so suggestions are exactly what we want. I'll check out Typer but as it's built on click I imagine it'll have the same issue of coupling argparsing with everything else...
